### ISSUE TITLE
Setup CI/CD

### ISSUE THREAD

**Issue by jackzampolin** [cosmos/ethermint_bak#9]
setup packaging and builds early so that they are there later.

cc @greg-szabo

**Comment by greg-szabo** [comment]
Braindump:
1. Written agreement on how the `develop` and `master` branches are handled. (I propose that both build at all times and they only get PRs, not direct commits. They should be locked down to merge by appointed reviewers only.) - Should be in the coding practices docs.
2. Agree or discuss: unit testing is done in CircleCI
3. Agree or discuss: integration testing is done in CircleCI (using the localnet commands)
4. Agree or discuss: deployment is triggered on PR merges from `develop` and `master`. There is no testing done during the release process. The end of the `develop` release process is a linux binary uploaded to S3 and deployed into a QA environment. (Need to agree on how to handle breaking changes in QA.) The end of the `master` release process is a new release on GitHub containing a link to the Changelog and the binaries built as well as a docker image uploaded to DockerHub.
5. Discuss: responsibilities. (Most should be on the dev team; providing tooling should be on the sre team.) I can hash this out more. Oh, I guess there's no "localnet" command for ethermint yet, either.
6. Discuss versioning: release management can do automatic version but that's in conflict with the manual semantic versioning set up in the `version.go` file.

**Comment by jackzampolin** [comment]
1. My inclination here is to follow the standards in `cosmos-sdk`. `master` is only working, shipped code while `develop` is working, unshipped code. Feature branches in `<name>/<feature>` format. I'm working on adapting the `CONTRIBUITING.md` from `cosmos-sdk` over here.
2. I would agree with this interested in hearing what @alexanderbez has to say.
3. I would agree with this. Might be difficult to setup as running integration test might require pulling copy of ETH blockchain and running the built binary on top of it. Don't know if we will be able to set this up in circle, might require some other automation...
4. I think initially we can hold off on QA environment for a bit, but I would like docker images and linux binaries sooner rather than later. Again, would be interested in what @alexanderbez and @AlexeyAkhunov think about this.
5. SRE only responsible for infra that supports builds/tests not for this repo. I can handle the deployment scripts and the build stuff here if thats cool with @alexanderbez and @AlexeyAkhunov 
6. Would be interested to here more thoughts on this from folks.

**Comment by greg-szabo** [comment]
1. cool, agreed
2. cool, agreed, depending on @alexanderbez 
3. Can we save a snapshot of the Ethereum blockchain and feed it to `localnet`? This might become a problem even when testing locally. The point here is to try to reuse the code we use for local testing. Details of how it's implemented can be broken out of here. The importance for me is to not go down the rabbit hole of bash scripts, like cosmos/tendermint did.
4. cool, agreed, depending on @alexanderbez & @AlexeyAkhunov . Let's nag ourselves with QA environment now and then. Let's focus on the rest for now.
5.  cool, agreed. The point here is that if something breaks, the devs should look at it first (assuming the infra was handed over and mature). There will be hand-holding at the beginning so dev team can get accustomed to it.
6. agreed. For example the current offering for release infra is Jenkins. There's good and bad in there but if people would prefer another self-hosted CI, we can discuss it now. (It's extra overhead so there should be a good reason for it.) Self-hosted is important because it will have access to the repo, which we don't want to give to third-parties, like CircleCI. Alternatively, I can offer intro to Jenkins courses.

**Comment by alexanderbez** [comment]
Wow, thanks for all the helpful input guys, yeah:

1. Agreed (preferably `master` should never be broken and be considered the canonical stable branch of the repo).
2. Agreed, Circle is just fine.
3. This might be interesting, but I agree with you @greg-szabo. We may need a more sophisticated setup for integration testing, but maybe Circle will suffice for now.
4. Agreed -- I'd shelf any QA stuff for now. I think the important thing is having releases and builds being triggered off of `master` just as @greg-szabo pointed out.
5. I'm cool with what you're suggesting just as long as decisions are discussed within the entire team (i.e. no surprises on the SRE front).
6. For releases, I would think it'd be wise to enforce semver, where the version is bumped (wherever applicable) in the actual release branch/PR. Then once that is merged into `master` an actual release binary can be built from that. Let's discuss further @greg-szabo

**Comment by greg-szabo** [comment]
6. That's how it's done in tendermint (and I guess cosmos-sdk too, but I'm not sure). Check the version package in either to see how it's implemented.

I have a better idea for automation that doesn't rely on a developer bumping the release number (major and minor should suffice for that). I'll show it in my f11 repo (I can demo soon). It applies whatever build number the CI automation comes up with and uses that as release number. This ensures that the build number is always higher than before and lets the developer manage their feature/bugfix releases in major/minor. This resolves the issue when the developer forgets to bump the release number and the release process fails with "existing tag".

It might be useful only after release 1.0 if we want to keep doing lots of breaking changes in 0.x.
